### PR TITLE
switch Version/VersionReq usages to RustcVersion 

### PR DESCRIPTION
**Pull request by Suyash458**
add `rustc-semver` to dependencies
switch `Version/VersionReq` usages to `RustcVersion`
changelog: none

### CODE REVIEW

**Comment by flip1995** [comment]
@rust-lang/clippy This PR introduces a new dependency: [`rustc-semver`](https://crates.io/crates/rustc-semver). Because this crate was written by me, probably someone else should also take a look at this.

`rustc-semver` is a minimalistic crate, that lets you parse versions of the Rust compiler and compare them. The crate is a `no_std` library with 100% code coverage, checked with the new [`source-based-code-coverage`](https://doc.rust-lang.org/nightly/unstable-book/compiler-flags/source-based-code-coverage.html).

**Comment by flip1995** [comment]
https://github.com/rust-lang/rust-clippy/blob/428ce12f6cf37f04e6e20cc959ff24b71c732911/tests/ui/min_rust_version_multiple_inner_attr.rs#L3

So the multiple inner attrs error has precedence over parsing errors.

File: clippy_lints/Cargo.toml [hunk 1]
```diff
@@ -28,6 +28,7 @@ smallvec = { version = "1", features = ["union"] }
 toml = "0.5.3"
 unicode-normalization = "0.1"
 semver = "0.11"
```
Comment (oli-obk) [inline clippy_lints/Cargo.toml:30]: can we remove the dependency on `semver` now?

**Review by oli-obk — COMMENTED** [review]
seems fine to me to switch to rustc-semver

File: clippy_lints/Cargo.toml (hunk 1 above)
Comment (flip1995) [reply clippy_lints/Cargo.toml:30]: I don't think so. It is still used in the ``wildcard_dependencies`` cargo lint

**Comment by Manishearth** [comment]
rustc-semver seems fine to me

File: clippy_lints/src/lib.rs [hunk 2]
```diff
@@ -966,22 +966,17 @@ pub fn register_plugins(store: &mut rustc_lint::LintStore, sess: &Session, conf:
...
-    store.register_late_pass(move || box matches::Matches::new(msrv.clone()));
-    let msrv = parsed_msrv.clone();
-    store.register_early_pass(move || box manual_non_exhaustive::ManualNonExhaustive::new(msrv.clone()));
-    let msrv = parsed_msrv;
-    store.register_late_pass(move || box manual_strip::ManualStrip::new(msrv.clone()));
-
+    store.register_late_pass(move || box methods::Methods::new(msrv));
+    store.register_late_pass(move || box matches::Matches::new(msrv));
+    store.register_early_pass(move || box manual_non_exhaustive::ManualNonExhaustive::new(msrv));
+    store.register_late_pass(move || box manual_strip::ManualStrip::new(msrv));
```
Comment (Suyash458) [inline clippy_lints/src/lib.rs:979]: Don't think we need to `clone()` unnecessarily anymore

File: clippy_lints/src/utils/mod.rs [hunk 3]
```diff
@@ -75,8 +75,8 @@ pub fn parse_msrv(msrv: &str, sess: Option<&Session>, span: Option<Span>) -> Opt
     None
 }
 
-pub fn meets_msrv(msrv: Option<&VersionReq>, lint_msrv: &Version) -> bool {
-    msrv.map_or(true, |msrv| !msrv.matches(lint_msrv))
+pub fn meets_msrv(msrv: Option<&RustcVersion>, lint_msrv: &RustcVersion) -> bool {
+    msrv.map_or(true, |msrv| msrv >= lint_msrv)
```
Comment (Suyash458) [inline clippy_lints/src/utils/mod.rs:79]: @flip1995 should we add a test to check equality of `msrv` as well?

Comment (flip1995) [reply clippy_lints/src/utils/mod.rs:79]: Yeah a test for equality and for just in and just out would be good. You can put them in the same test file by using `mod { #![msrv] }`

File: clippy_lints/src/lib.rs (hunk 2 above)
Comment (flip1995) [reply clippy_lints/src/lib.rs:979]: Oh yeah, good catch. `RustcVersion` is `Copy`.

Comment (flip1995) [reply clippy_lints/src/lib.rs:979]: Oh yeah, good catch. `RustcVersion` is `Copy`.

File: clippy_lints/src/utils/mod.rs (hunk 3 above)
Comment (flip1995) [inline clippy_lints/src/utils/mod.rs:79]: I think we should use the newly implemented `meets` version here. This change should not have an influence on the tests.

```suggestion
    msrv.map_or(true, |msrv| msrv.meets(*lint_msrv))
```

That way our implementation of MSRV is conform with the definition in the [RFC](https://github.com/rust-lang/rfcs/blob/master/text/2523-cfg-path-version.md#cfgversionsemver)

**Review by flip1995 — CHANGES_REQUESTED** [review]

File: clippy_lints/src/utils/mod.rs (hunk 3 above)
Comment (Suyash458) [reply clippy_lints/src/utils/mod.rs:79]: Makes sense

**Review by flip1995 — APPROVED** [review]
Waiting on rustup

**Comment by Suyash458** [comment]
@flip1995 does this need a rebase?

**Comment by flip1995** [comment]
Nope, we're waiting for a rustup (which I already opened a PR for). The problem is that the rustup will need a ICE fix in rustc.

If a issue is labeled as <label>`S-waiting-on-rustup`</label> nothing more has to be done by the contributor.

**Pull request merged**